### DISCUSSION TITLE
How can I build a statically linked nu for x86_64-unknown-linux-musl?

### DISCUSSION THREAD

**Discussion by theoparis** [nushell/nushell#13516] [Q&A]
I tried: `cargo install nu --target x86_64-unknown-linux-musl --no-default-features --features static-link-openssl` and got linker errors as seen below 😕 

```
  = note: ld.lld: error: undefined symbol: __isoc23_strtol
          >>> referenced by rand_unix.c
          >>>               libdefault-lib-rand_unix.o:(ossl_pool_acquire_entropy) in archive /tmp/rustcTTkUQa/libopenssl_sys-3fbb1e84b37279c7.rlib
          >>> referenced by rand_unix.c
          >>>               libdefault-lib-rand_unix.o:(ossl_pool_acquire_entropy) in archive /tmp/rustcTTkUQa/libopenssl_sys-3fbb1e84b37279c7.rlib
          
          ld.lld: error: undefined symbol: fcntl64
          >>> referenced by sqlite3.c
          >>>               0343851c9008f270-sqlite3.o:(aSyscall) in archive /tmp/rustcTTkUQa/liblibsqlite3_sys-fe0d47b99c71a1ac.rlib
          clang: error: linker command failed with exit code 1 (use -v to see invocation)
          

error: could not compile `nu` (bin "nu") due to 1 previous error
error: failed to compile `nu v0.96.1`, intermediate artifacts can be found at `/tmp/cargo-installTmUrao`.
To reuse those artifacts with a future compilation, set the environment variable `CARGO_TARGET_DIR` to that path.
```

Edit: `podman run --rm -it docker.io/instrumentisto/rust:nightly-alpine` + `apk add musl-dev perl make` + the initial cargo install command along with adding crt-static to rustflags seems to work. Not sure why my host system lld fails though

**Comment by ghost** [comment]
im not a c expert, but afaik static linking requires the exact libraries to be present on the system. if you try to compile it on a GNU-linux with MUSL-linux as target the MUSL libraries (in this case `fcntl64`) are missing (or a GNU version is at their location).  
alpine uses MUSL and within a container the container-OS libraries get used.

if this doesn't explain the issue it might be helpful (for others) to know which OS, etc you use. the output of `uname -a` (or `^uname -a` within nushell) would provide all this.

**Comment by chrda81** [comment]
I had to build nushell for target "i686-unknown-linux-musl" and stumbled over https://github.com/cross-rs/cross. I had to:

1) Install cross (see additional requirements on page above):

```sh
cargo install cross --git https://github.com/cross-rs/cross
```

2) Create a Cross.toml file in nushell's root directory:

```toml
# Configuration for cross-rs: https://github.com/cross-rs/cross
# Run cross-rs like this:
# cross build --target i686-unknown-linux-musl --release --features=static-link-openssl

#  NOTE: for musl you will need to build with --features=static-link-openssl
[target.i686-unknown-linux-musl]
pre-build = [
    "dpkg --add-architecture i386",
    "apt-get update && apt-get install --assume-yes libssl-dev:i386 clang"
]
```

3) Build everything with

```sh
cross build --target i686-unknown-linux-musl --release --features=static-link-openssl
```

Hope that helps.
